rollout/sync.go

```diff
@@ -584,7 +584,20 @@ func (c *rolloutContext) calculateRolloutConditions(newStatus v1alpha1.RolloutSt
 	} else {
 		if completeCond != nil {
 			updateCompletedCond := conditions.NewRolloutCondition(v1alpha1.RolloutCompleted, corev1.ConditionFalse, conditions.RolloutCompletedReason, conditions.RolloutCompletedReason)
-			conditions.SetRolloutCondition(&newStatus, *updateCompletedCond)
+			changed := conditions.SetRolloutCondition(&newStatus, *updateCompletedCond)
+
+			// When a fully promoted rollout becomes Incomplete, e.g., due to the ReplicaSet status changes like
+			// pod restarts, evicted -> recreated, we'll need to reset the rollout's condition to `PROGRESSING` to
+			// avoid any timeouts.
+			if changed {
+				existProgressingCondition := conditions.GetRolloutCondition(newStatus, v1alpha1.RolloutProgressing)
+				if existProgressingCondition != nil {
+					conditions.RemoveRolloutCondition(&newStatus, v1alpha1.RolloutProgressing)
+				}
+				c.log.Infof(conditions.RolloutBecomesIncompleteMessage)
+				newProgressingCondition := conditions.NewRolloutCondition(v1alpha1.RolloutProgressing, corev1.ConditionTrue, conditions.RolloutBecomesIncomplete, conditions.RolloutBecomesIncompleteMessage)
+				conditions.SetRolloutCondition(&newStatus, *newProgressingCondition)
```

Comment:
@huikang i think this message and reason will get clobbered by the switch statement that comes later.

Instead of updating the Progressing condition here (where we detect that we became incomplete), can we update the Progressing condition at the current place in the code where we are managing the Progressing condition (in the switch statement)? This will require some new boolean (e.g. becameIncomplete) set in the completed check:

```golang
	var becameIncomplete bool // remember if we transitioned from completed:
	if !isPaused && conditions.RolloutComplete(c.rollout, &newStatus) {
...
	} else {
		if completeCond != nil {
			updateCompletedCond := conditions.NewRolloutCondition(v1alpha1.RolloutCompleted, corev1.ConditionFalse, conditions.RolloutCompletedReason, conditions.RolloutCompletedReason)
			becameIncomplete = conditions.SetRolloutCondition(&newStatus, *updateCompletedCond)
changed := conditions.SetRolloutCondition(&newStatus, *updateCompletedCond)
...

	if !isCompleteRollout && !isAborted {
		switch {
...
		case conditions.RolloutProgressing(c.rollout, &newStatus):
...
			// Give a more accurate reason for the Progressing condition
			if newStatus.StableRS == newStatus.CurrentPodHash {
				reason = conditions.ReplicaSetNotAvailableReason
				msg = conditions.NotAvailableMessage // re-use existing message
			} else {
				reason = conditions.ReplicaSetUpdatedReason
			}
			condition := conditions.NewRolloutCondition(v1alpha1.RolloutProgressing, corev1.ConditionTrue, reason, msg)

			if currentCond != nil  || becameIncomplete {
				if currentCond.Status == corev1.ConditionTrue {
					condition.LastTransitionTime = currentCond.LastTransitionTime
				}
				conditions.RemoveRolloutCondition(&newStatus, v1alpha1.RolloutProgressing)
			}
			conditions.SetRolloutCondition(&newStatus, *condition)
```

Reply:
Thanks for the suggestion. Separating the processing of progressing and complete condition sounds better. Let me update the PR to test the logic.